---
title: "Day 3 - Module 09 Heatmap"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            markdown: GFM
            code_download: true
---
<style type="text/css">
body{ /* Normal  */
      font-size: 14pt;
  }
pre {
  font-size: 12pt
}
code.r{
  font-size: 12pt;
}
</style>

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

<br>

# [Breakout Exercise 2] - Finding annotations

<br>


```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
load("rdata/RunningData.RData")
```


**15 Minutes**

<br>



<br>

### Instructions:

<br>

- One group member should share their screen in the breakout room. If nobody volunteers, a helper may randomly select someone.
- The group members should discuss the exercise and work together to find a solution.
- If there is time after a solution is found, allow all members to complete the exercise.

<br>

- First, ...


> Hint:
> 
>

<br>


## [Breakout Exercise 2]: Build a BioMart query and annotate the DE results

Next,[build your query](https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#how-to-build-a-biomart-query) (using the linked help page) to:   

1. Retrieve the information (aka: attributes) we want to add to our DE results table (e.g. `ensembl_gene_id`, `external_gene_name`)...    
1. Using the `getBM` function (*Hint: use the `?` operator to see documentation for the function*)...
1. Filtering on the ensembl gene id attribute...    
1. With values that are the ENSEMBL ids from our experiment (*Hint: the row.names of the
  `dds` assay)...    
1. And store it in an object named `GeneKey`.



Add this to the DE results table `res_WT` and store the new table in an object named `res_WT_anno`. 
*Hint: look at the documentation for either the `merge` function or the tidyverse `join` function and make sure to add a column to hold the gene ids prior to adding the annotations.

<details>
    <summary>*Click for solution for adding annotations*</summary>
    First, create a new table called `res_WT_anno` that includes a column with the ENSEMBL ids named `genes` using the `mutate` function. Then use the `left_join` function to combine the `GeneKey` table with the `res_WT` DE results.
```{r updateDETableTidyr}
res_WT_anno <- as.data.frame(res_WT) %>%
  mutate(genes = row.names(res_WT)) %>%
  left_join(GeneKey, by =c("genes" = "ensembl_gene_id")) %>%
  relocate(c("genes", "external_gene_name")) # optionally, re-order columns to make output more readable

head(res_WT_anno)
```
    Alternatively, if you are more familiar with base functions:
```{r updateDETableBase}
res_WT_anno <- res_WT # copy table
res_WT_anno <- cbind(genes=row.names(res_WT_anno), res_WT_anno[ ,c(1:6)])
res_WT_anno <- as.data.frame(res_WT_anno)

# combine the two tables using the merge function (similar to join from `tidyverse`)
res_WT_anno <- merge(GeneKey, res_WT_anno, by.x = "ensembl_gene_id", by.y="genes", all.x = FALSE, all.y = TRUE)

head(res_WT_anno)
```
    Notice that not all genes were annotated with an ENSEMBl gene id or gene description. While we are able to annotate our results, we should be very cautious as the gene symbol is not a good unique identifier plus we did not use a UCSC annotation resource so the HUGO gene symbol may not always match. However, this code is similar to the steps needed to annotate ENSEMBL id based results, like what would have been generated from yesterday's alignments, with more interpretable gene symbols.

</details>

  
```{r WriteOut.RData, eval=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
save.image(file = "rdata/RunningData.RData")
```


